### PR TITLE
stateDB txlog store enhancement

### PR DESCRIPTION
**Pull request by JayT106**
## Description

after profiling BenchmarkEmitLogs in #436
found we can improve the emit log speed by changing the log store structure.

Therefore, we store each log separately `prefixlogkey | txHash | log.Index`, not stores by []logs 

The change improves the log store by around 90x within the testcase.

In #436:
```
goos: darwin
goarch: amd64
pkg: github.com/tharsis/ethermint/x/evm/keeper
cpu: Intel(R) Core(TM) i7-9750H CPU @ 2.60GHz
BenchmarkEmitLogs-12    	       5	1197719670 ns/op
```

In This PR:
```
goos: darwin
goarch: amd64
pkg: github.com/tharsis/ethermint/x/evm/keeper
cpu: Intel(R) Core(TM) i7-9750H CPU @ 2.60GHz
BenchmarkEmitLogs-12    	     453	  13216579 ns/op
```

Updated PR:
```
pkg: github.com/tharsis/ethermint/x/evm/keeper
cpu: Intel(R) Core(TM) i7-9750H CPU @ 2.60GHz
BenchmarkEmitLogs
BenchmarkEmitLogs-12         	      97	  12626105 ns/op
```
______

For contributor use:

- [x] Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#pr-targeting))
- [x] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/master/docs/building-modules/structure.md).
- [ ] Wrote unit and integration [tests](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#testing)
- [ ] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`)
- [ ] Added  relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [ ] Added a relevant changelog entry to the `Unreleased` section in `CHANGELOG.md`
- [ ] Re-reviewed `Files changed` in the Github PR explorer

______

For admin use:

- [ ] Added appropriate labels to PR (ex. `WIP`, `R4R`, `docs`, etc)
- [ ] Reviewers assigned
- [ ] Squashed all commits, uses message "Merge pull request #XYZ: [title]" ([coding standards](https://github.com/tendermint/coding/blob/master/README.md#merging-a-pr))

### CODE REVIEW

File: x/evm/keeper/statedb.go [hunk 1]
```diff
@@ -608,10 +608,10 @@ func (k *Keeper) AddLog(log *ethtypes.Log) {
 	log.Index = uint(k.GetLogSizeTransient())
 	k.IncreaseLogSizeTransient()
 
-	logs := k.GetTxLogs(log.TxHash)
-	logs = append(logs, log)
-
-	k.SetLogs(log.TxHash, logs)
+	s := prefix.NewStore(k.Ctx().KVStore(k.storeKey), types.KeyPrefixTxLogNum)
+	bi := new(big.Int).SetBytes(s.Get(log.TxHash.Bytes()))
+	s.Set(log.TxHash.Bytes(), new(big.Int).SetInt64(bi.Int64()+1).Bytes())
```
Comment (fedekunze) [inline x/evm/keeper/statedb.go:613]: curious on why are you using the big.Int to set the hash?

File: x/evm/keeper/keeper.go [hunk 2]
```diff
@@ -229,50 +229,124 @@ func (k Keeper) ResetRefundTransient(ctx sdk.Context) {
...
 
 // GetLogs returns the current logs for a given transaction hash from the KVStore.
 // This function returns an empty, non-nil slice if no logs are found.
 func (k Keeper) GetTxLogs(txHash common.Hash) []*ethtypes.Log {
-	store := prefix.NewStore(k.Ctx().KVStore(k.storeKey), types.KeyPrefixLogs)
+	store := prefix.NewStore(k.Ctx().KVStore(k.storeKey), types.KeyPrefixTxLogNum)
 
 	bz := store.Get(txHash.Bytes())
-	if len(bz) == 0 {
+	if bz == nil {
```
Comment (fedekunze) [inline x/evm/keeper/keeper.go:272]:
```suggestion
	if len(bz) == 0 {
```

File: x/evm/keeper/keeper.go [hunk 3]
```diff
@@ -229,50 +229,124 @@ func (k Keeper) ResetRefundTransient(ctx sdk.Context) {
...
+		return []*ethtypes.Log{}
+	}
+
+	logNum := new(big.Int).SetBytes(bz)
```
Comment (fedekunze) [inline x/evm/keeper/keeper.go:276]: ditto, why big.Int?

File: x/evm/keeper/keeper.go [hunk 4]
```diff
@@ -229,50 +229,124 @@ func (k Keeper) ResetRefundTransient(ctx sdk.Context) {
...
+	}
+	return logs
 }
 
 // SetLogs sets the logs for a transaction in the KVStore.
 func (k Keeper) SetLogs(txHash common.Hash, logs []*ethtypes.Log) {
-	store := prefix.NewStore(k.Ctx().KVStore(k.storeKey), types.KeyPrefixLogs)
+	var s prefix.Store
+	len := len(logs)
+	if len > 0 {
```
Comment (fedekunze) [inline x/evm/keeper/keeper.go:307]: why wouldn’t this be the case?

File: x/evm/keeper/grpc_query.go [hunk 5]
```diff
@@ -232,15 +232,19 @@ func (k Keeper) BlockLogs(c context.Context, req *types.QueryBlockLogsRequest) (
...
-		var txLog types.TransactionLogs
+		var txLog types.Log
 		k.cdc.MustUnmarshal(value, &txLog)
 
-		if len(txLog.Logs) > 0 && txLog.Logs[0].BlockHash == req.Hash {
+		if txLog.BlockHash == req.Hash {
 			if accumulate {
-				txLogs = append(txLogs, txLog)
+				if logs[txLog.TxHash] == nil {
+					logs[txLog.TxHash] = []*types.Log{}
```
Comment (leejw51crypto) [inline x/evm/keeper/grpc_query.go:244]: 2 ifs can be one if?

File: x/evm/keeper/keeper.go [hunk 6]
```diff
@@ -229,50 +229,124 @@ func (k Keeper) ResetRefundTransient(ctx sdk.Context) {
...
 	store := ctx.KVStore(k.storeKey)
-	iterator := sdk.KVStorePrefixIterator(store, types.KeyPrefixLogs)
+	iterator := sdk.KVStorePrefixIterator(store, types.KeyPrefixTxLogNum)
 	defer iterator.Close()
 
 	txsLogs := []types.TransactionLogs{}
 	for ; iterator.Valid(); iterator.Next() {
-		var txLog types.TransactionLogs
-		k.cdc.MustUnmarshal(iterator.Value(), &txLog)
+		logNum := new(big.Int).SetBytes(iterator.Value())
```
Comment (leejw51crypto) [inline x/evm/keeper/keeper.go:237]: is this logCount ? or logID?

File: x/evm/keeper/keeper.go (hunk 3 above)
Comment (leejw51crypto) [reply x/evm/keeper/keeper.go:276]: you mean int64?

File: x/evm/keeper/statedb.go (hunk 1 above)
Comment (leejw51crypto) [reply x/evm/keeper/statedb.go:613]: just bytes would be simpler

File: x/evm/keeper/keeper.go [hunk 7]
```diff
@@ -229,50 +229,124 @@ func (k Keeper) ResetRefundTransient(ctx sdk.Context) {
...
+	if b == nil {
+		return
+	}
+
+	txLogCnt := new(big.Int).SetBytes(b)
+
 	store.Delete(txHash.Bytes())
+
+	store = prefix.NewStore(ctx.KVStore(k.storeKey), types.KeyPrefixLogs)
+	for i := int64(0); i < txLogCnt.Int64(); i++ {
```
Comment (leejw51crypto) [inline x/evm/keeper/keeper.go:343]: how about txLogCount

File: x/evm/types/key.go [hunk 8]
```diff
@@ -30,6 +30,7 @@ const (
 	prefixLogs
 	prefixCode
 	prefixStorage
+	prefixTxLogNum
```
Comment (leejw51crypto) [inline x/evm/types/key.go:33]: how about
prefixTxLogNum , it's a little unclear whether it's ID or count.

prefixTxLogCount ?

File: x/evm/keeper/statedb.go (hunk 1 above)
Comment (JayT106) [reply x/evm/keeper/statedb.go:613]: > curious on why are you using the big.Int to set the hash?

Mainly want to bank the space and 2 liners,  so most of the cases will store only 1 byte, I think fewer cases might need to use 2 bytes(more than 255 logs in a transaction)

I can add a helper to deal with the dynamic bytes transfer, or do you think we can always use 2 bytes to store the transaction log count?

File: x/evm/keeper/keeper.go (hunk 3 above)
Comment (JayT106) [reply x/evm/keeper/keeper.go:276]: https://github.com/tharsis/ethermint/pull/461#discussion_r693009503

File: x/evm/keeper/keeper.go (hunk 6 above)
Comment (JayT106) [reply x/evm/keeper/keeper.go:237]: yes, it's logCount, I will change it.

File: x/evm/keeper/keeper.go (hunk 7 above)
Comment (JayT106) [reply x/evm/keeper/keeper.go:343]: sure!

File: x/evm/types/key.go (hunk 8 above)
Comment (JayT106) [reply x/evm/types/key.go:33]: sure

File: x/evm/keeper/keeper.go (hunk 4 above)
Comment (JayT106) [reply x/evm/keeper/keeper.go:307]: Do we confirm the `lens` of arguments `logs` will always be greater than 0?
I was thinking if the logs are empty, so we don't need to store anything.

**Review by fedekunze — APPROVED** [review]
ACK. Thanks @JayT106! Can you add a Changelog entry under the improvements section 🙏

**Comment by JayT106** [comment]
> ACK. Thanks @JayT106! Can you add a Changelog entry under the improvements section 🙏

@fedekunze Hi, thanks for the review. The code likes like has been merged. Should I open another PR for adding Changelog?

**Comment by JayT106** [comment]
> ACK. Thanks @JayT106! Can you add a Changelog entry under the improvements section 🙏

Sure, will do?

File: CHANGELOG.md [hunk 9]
```diff
@@ -54,6 +54,9 @@ Ref: https://keepachangelog.com/en/1.0.0/
 * (encoding) [tharsis#478](https://github.com/tharsis/ethermint/pull/478) Register `Evidence` to amino codec.
 * (rpc) [tharsis#478](https://github.com/tharsis/ethermint/pull/481) Getting the node configuration when calling the `miner` rpc methods.
 
+### Improvements
+* (evm) [tharsis#461](https://github.com/tharsis/ethermint/issues/461) StateDB transaction log stores enhancement.
```
Comment (fedekunze) [inline CHANGELOG.md:58]:
```suggestion

* (evm) [tharsis#461](https://github.com/tharsis/ethermint/pull/461) Increase performance of `StateDB` transaction log storage (r/w).
```

**Pull request merged**